src/platform/GenericOTARequestorDriver.cpp

```diff
@@ -94,29 +162,158 @@ void GenericOTARequestorDriver::UpdateSuspended(System::Clock::Seconds32 delay)
+    CancelDelayedAction([](System::Layer *, void * context) { ToDriver(context)->mImageProcessor->Apply(); }, this);
+    CancelDelayedAction([](System::Layer *, void * context) { ToDriver(context)->mRequestor->ApplyUpdate(); }, this);
 }
 
+    // SL TODO: Get rid of state
 void GenericOTARequestorDriver::ScheduleDelayedAction(UpdateFailureState state, System::Clock::Seconds32 delay,
-                                                      System::TimerCompleteCallback action)
+                                                      System::TimerCompleteCallback action, void * aAppState)
+{
+    SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState);
+
+    //    if (error != CHIP_NO_ERROR)
+    // {
+    //    HandleError(state, error);
+    // }
+}
+
+void GenericOTARequestorDriver::CancelDelayedAction(System::TimerCompleteCallback action, void * aAppState)
+{
+    SystemLayer().CancelTimer(action, aAppState);
+}
+
+void GenericOTARequestorDriver::ProcessAnnounceOTAProviders(const ProviderLocationType &providerLocation, 
+                                                          app::Clusters::OtaSoftwareUpdateRequestor::OTAAnnouncementReason announcementReason)
+{
+    // If reason is URGENT_UPDATE_AVAILABLE, we start OTA immediately. Otherwise, respect the timer value set in mOtaStartDelayMs.
+    // This is done to exemplify what a real-world OTA Requestor might do while also being configurable enough to use as a test app.
+    uint32_t msToStart = 0;
+    switch (announcementReason)
+    {
+    case OTAAnnouncementReason::kSimpleAnnouncement:
+    case OTAAnnouncementReason::kUpdateAvailable:
+        msToStart = mOtaStartDelayMs;
+        break;
+    case OTAAnnouncementReason::kUrgentUpdateAvailable:
+        msToStart = kImmediateStartDelayMs;
+        break;
+    default:
+        ChipLogError(SoftwareUpdate, "Unexpected announcementReason: %u", static_cast<uint8_t>(announcementReason));
+        return;
+    }
+
+    // IMPLEMENTATION CHOICE:
+    // This implementation of the OTARequestor driver ignores the announcement if an update is in progress, 
+    // otherwise it queries the provider passed in the announcement
+
+    if(mRequestor->GetCurrentUpdateState() != OTAUpdateStateEnum::kIdle) {
+        ChipLogProgress(SoftwareUpdate, "State is not kIdle, ignoring the AnnounceOTAProviders. State: %d", (int)mRequestor->GetCurrentUpdateState());
+        return;
+   }
+
+    // Point the OTARequestor to the announced provider 
+    mRequestor->SetCurrentProviderLocation(providerLocation);
+    mLastUsedProvider = providerLocation;
+
+    ScheduleDelayedAction(UpdateFailureState::kQuerying, System::Clock::Seconds32(msToStart/1000), StartDelayTimerHandler, mRequestor);
+}
+
+void GenericOTARequestorDriver::DriverSendQuery()
+{
+
+    // IMPLEMENTATION CHOICE
+    // In this implementation explicitly triggering a query cancels any in-progress update.
+    UpdateCancelled();
```

Comment:
Do we want this to happen? This is saying that a periodic query trumps all existing in progress updates?

Reply:
I think this could be helpful as a catch all for cases like https://github.com/project-chip/connectedhomeip/issues/15423 but in particular, the periodic query is set to 1 min in the Linux main func which makes the constant cancelling of existing in-progress update bad, IMO.

Reply:
Yes, I think this is reasonable. Periodic query should not be happening if there is an update in progress (we stop the timer as soon as we start the process). This makes sure we are in a consistent state. 
For application-triggered query we'll cancel anything that is in progress currently -- I think this is reasonable (the alternative would be to return an error and then have the application retry)

Reply:
I'm fine with this being an implementation choice. Just wanted to keep in mind that if the application doesn't have any guards against constantly triggering (user keeps pressing "Check for updates" button), then the OTA update will keep getting interrupted. Where as if we denied application triggers, the worst that can happen is the OTA will eventually finish and the user got what they wanted anyway.

Another thought on this could be to eventually have `TriggerImmediateQuery` return the current state so the caller can inform that the error is due to OTA in progress. Could be a nice to have later.